mempool/src/interface/mempool_interface_impl.rs

```diff
@@ -13,100 +13,141 @@
-
-pub struct MempoolInterfaceImpl {
-    sender: mpsc::UnboundedSender<MempoolMethodCall>,
+struct MempoolInterfaceImpl<M> {
+    pool: Mempool<M>,
 }
 
-impl MempoolInterfaceImpl {
-    pub fn new<M: GetMemoryUsage + Sync + Send + 'static>(
+impl<M: GetMemoryUsage + Sync + Send + 'static> MempoolInterfaceImpl<M> {
+    pub fn new(
         chain_config: Arc<ChainConfig>,
         chainstate_handle: subsystem::Handle<Box<dyn ChainstateInterface>>,
         time_getter: TimeGetter,
         memory_usage_estimator: M,
-    ) -> Result<Self, crate::error::Error> {
-        let (sender, receiver) = tokio::sync::mpsc::unbounded_channel();
-
-        Mempool::new(
+    ) -> Self {
+        let pool = Mempool::new(
             chain_config,
             chainstate_handle,
             time_getter,
             memory_usage_estimator,
-            receiver,
-        )
-        .run()?;
+        );
+        Self { pool }
+    }
+
+    pub async fn subscribe_to_chainstate_events(
+        &mut self,
+    ) -> crate::Result<mpsc::UnboundedReceiver<(Id<Block>, BlockHeight)>> {
+        let (tx, rx) = mpsc::unbounded_channel();
+        let subscribe_func =
+            Arc::new(
+                move |chainstate_event: chainstate::ChainstateEvent| match chainstate_event {
+                    chainstate::ChainstateEvent::NewTip(block_id, block_height) => {
+                        log::info!(
+                            "Received a new tip with block id {:?} and block height {:?}",
+                            block_id,
+                            block_height
+                        );
+                        if let Err(e) = tx.send((block_id, block_height)) {
+                            log::error!("Mempool Event Handler closed: {:?}", e)
+                        }
+                    }
+                },
+            );
 
-        Ok(Self { sender })
+        self.pool
+            .chainstate_handle()
+            .call_mut(|this| this.subscribe_to_events(subscribe_func))
+            .await
+            .map_err(|_| crate::error::Error::SubsystemFailure)?;
+        Ok(rx)
     }
 }
 
 #[async_trait::async_trait]
-impl MempoolInterface for MempoolInterfaceImpl {
+impl<M: GetMemoryUsage + Sync + Send + 'static> MempoolSubsystemInterface
+    for MempoolInterfaceImpl<M>
+{
+    async fn run(
+        mut self,
+        mut call_rq: CallRequest<dyn MempoolInterface>,
+        mut shut_rq: ShutdownRequest,
+    ) {
+        let mut chainstate_events_rx = self
+            .subscribe_to_chainstate_events()
+            .await
+            .log_err()
+            .expect("chainstate event subscription");
+        loop {
+            tokio::select! {
+                () = shut_rq.recv() => break,
+                call = call_rq.recv() => call(&mut self).await,
+                Some((block_id, block_height)) = chainstate_events_rx.recv() => {
+                    self.pool.new_tip_set(block_id, block_height);
+                }
+            }
+        }
```

Comment:
As per call: This code mixes mempool-specific code with code that handles calls and shutdown. The call and shutdown handling is normally provided by the subsystem framework but is unavailable if custom event loop is used. It would be preferable for the event loop + call handling + shutdown handling to be still provided by the subsystem framework while exposing extension points to give each subsystem the ability to react to inputs other than calls and shutdown requests. This is a matter for subsequent research.